controller_manager/test/test_controller_manager_srvs.cpp

```diff
@@ -480,3 +480,574 @@ TEST_F(TestControllerManagerSrvs, configure_controller_srv)
+  cm_->add_controller(
+    test_chained_controller_2, TEST_CHAINED_CONTROLLER_2,
+    test_chainable_controller::TEST_CONTROLLER_CLASS_NAME);
+  cm_->add_controller(
+    test_controller, test_controller::TEST_CONTROLLER_NAME,
+    test_controller::TEST_CONTROLLER_CLASS_NAME);
+  // get controller list before configure
+  auto result = call_service_and_wait(*client, request, srv_executor);
+  // check chainable controller
+  ASSERT_EQ(6u, result->controller.size());
+  ASSERT_EQ(result->controller[0].name, TEST_CHAINED_CONTROLLER_4);
+  ASSERT_EQ(result->controller[1].name, TEST_CHAINED_CONTROLLER_1);
+  ASSERT_EQ(result->controller[2].name, TEST_CHAINED_CONTROLLER_3);
+  ASSERT_EQ(result->controller[3].name, TEST_CHAINED_CONTROLLER_5);
+  ASSERT_EQ(result->controller[4].name, TEST_CHAINED_CONTROLLER_2);
+  // check test controller
+  ASSERT_EQ(result->controller[5].name, "test_controller_name");
+
+  // configure controllers
+  for (const auto & controller :
+       {TEST_CHAINED_CONTROLLER_4, TEST_CHAINED_CONTROLLER_3, TEST_CHAINED_CONTROLLER_5,
+        TEST_CHAINED_CONTROLLER_2, TEST_CHAINED_CONTROLLER_1,
+        test_controller::TEST_CONTROLLER_NAME})
+    cm_->configure_controller(controller);
+
+  // get controller list after configure
+  result = call_service_and_wait(*client, request, srv_executor);
+  ASSERT_EQ(6u, result->controller.size());
+
+  // activate controllers
+  cm_->switch_controller(
+    {TEST_CHAINED_CONTROLLER_1}, {},
+    controller_manager_msgs::srv::SwitchController::Request::STRICT, true, rclcpp::Duration(0, 0));
+  cm_->switch_controller(
+    {TEST_CHAINED_CONTROLLER_3, TEST_CHAINED_CONTROLLER_5, TEST_CHAINED_CONTROLLER_2,
+     TEST_CHAINED_CONTROLLER_4},
+    {}, controller_manager_msgs::srv::SwitchController::Request::STRICT, true,
+    rclcpp::Duration(0, 0));
+  cm_->switch_controller(
+    {test_controller::TEST_CONTROLLER_NAME}, {},
+    controller_manager_msgs::srv::SwitchController::Request::STRICT, true, rclcpp::Duration(0, 0));
+  // get controller list after activate
+  result = call_service_and_wait(*client, request, srv_executor);
+
+  ASSERT_EQ(6u, result->controller.size());
+  // reordered controllers
+  ASSERT_EQ(result->controller[0].name, "test_controller_name");
+  ASSERT_EQ(result->controller[1].name, TEST_CHAINED_CONTROLLER_5);
+  ASSERT_EQ(result->controller[2].name, TEST_CHAINED_CONTROLLER_4);
+  ASSERT_EQ(result->controller[3].name, TEST_CHAINED_CONTROLLER_3);
+  ASSERT_EQ(result->controller[4].name, TEST_CHAINED_CONTROLLER_2);
+  ASSERT_EQ(result->controller[5].name, TEST_CHAINED_CONTROLLER_1);
+  RCLCPP_ERROR(srv_node->get_logger(), "Check successful!");
+}
+
+TEST_F(TestControllerManagerSrvs, list_sorted_complex_chained_controllers)
+{
+  /// The simulated controller chain branching is:
+  /// test_controller_name -> chain_ctrl_7 -> chain_ctrl_6 -> chain_ctrl_2 -> chain_ctrl_1
+  /// &
+  /// chain_ctrl_6 -> chain_ctrl_5 -> chain_ctrl_4 -> chain_ctrl_3
```

Comment:
Can add here show actually controllers need reference interfaces of each other. This would help to have this clear for the future if we do some changes. Now you have a chain, but it's is hard to understand why without paper to write down how controllers are connected.

Reply:
Done added!